### PR TITLE
fix: a deadlock by refactoring listBuckets() under site replication

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
fix: a deadlock by refactoring listBuckets() under site replication

## Motivation and Context
fixes a deadlock

## How to test this PR?
Easy to reproduce take a disk offline you can see
listBuckets() would leak locks. 

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

File: cmd/erasure-sets.go [hunk 1]
```diff
@@ -874,45 +894,6 @@ func (s *erasureSets) ListBuckets(ctx context.Context, opts BucketOptions) (buck
...
-					return nil
-				}
-				return err
-			}
-			for _, volName := range volsInfo {
-				mu.Lock()
-				if _, ok := delBuckets[volName]; !ok {
-					vi, err := storageDisks[index].StatVol(ctx, pathJoin(minioMetaBucket, bucketMetaPrefix, deletedBucketsPrefix, volName))
-					if err != nil {
-						return err
```
Comment (harshavardhana) [inline cmd/erasure-sets.go:900]: The deadlock was here..

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :x: [more...](#66b243e6-5bed-4ab3-9b81-93b441c845b5) |
<a name="66b243e6-5bed-4ab3-9b81-93b441c845b5"></a>
### 16323-d8b2d7f/mint-compress-encrypt-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      15.15.15.6:31664
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp 6a5207472401:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 1 seconds
(2/14) Running aws-sdk-java tests ... done in 2 seconds
(3/14) Running aws-sdk-php tests ... done in 42 seconds
(4/14) Running aws-sdk-ruby tests ... done in 7 seconds
(5/14) Running awscli tests ... done in 1 minutes and 1 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 15 seconds
(8/14) Running minio-go tests ... done in 44 seconds
(9/14) Running minio-java tests ... FAILED in 26 seconds
{
  "name": "minio-java",
  "function": "addUser()",
  "duration": 384,
  "status": "FAIL",
  "error": "java.lang.RuntimeException: Request failed with response: {\"Code\":\"InternalError\",\"Message\":\"We encountered an internal error, please try again. (Server not initialized, please try again)\",\"Resource\":\"/minio/admin/v3/add-user\",\"RequestId\":\"1734F170632610C7\",\"HostId\":\"ccc303f1-f7ab-4ccf-8943-11866125b346\"}\n >>> [io.minio.admin.MinioAdminClient.execute(MinioAdminClient.java:206), io.minio.admin.MinioAdminClient.addUser(MinioAdminClient.java:237), TestMinioAdminClient.addUser(TestMinioAdminClient.java:47), TestMinioAdminClient.runAdminTests(TestMinioAdminClient.java:160), FunctionalTest.runTests(FunctionalTest.java:3782), FunctionalTest.runEndpointTests(FunctionalTest.java:3881), FunctionalTest.main(FunctionalTest.java:4002)]"
}
(9/14) Running minio-js tests ... done in 50 seconds
(10/14) Running minio-py tests ... done in 1 minutes and 39 seconds
(11/14) Running s3cmd tests ... done in 15 seconds
(12/14) Running s3select tests ... done in 3 seconds
(13/14) Running versioning tests ... done in 3 minutes and 5 seconds

Executed 13 out of 14 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-fs.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#07e8fc89-1259-4d7c-9e4e-9b31e3229aa4) |
| mint-compress-encrypt-dist-erasure.sh | :x: [more...](#d0a31f5a-b908-425e-9a60-a5764caaa615) |
<a name="d0a31f5a-b908-425e-9a60-a5764caaa615"></a>
### 16323-79822b3/mint-compress-encrypt-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      15.15.15.2:31649
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp 1e4254ad1158:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 1 seconds
(2/14) Running aws-sdk-java tests ... done in 2 seconds
(3/14) Running aws-sdk-php tests ... done in 42 seconds
(4/14) Running aws-sdk-ruby tests ... done in 7 seconds
(5/14) Running awscli tests ... done in 59 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 16 seconds
(8/14) Running minio-go tests ... done in 45 seconds
(9/14) Running minio-java tests ... FAILED in 27 seconds
{
  "name": "minio-java",
  "function": "addUser()",
  "duration": 406,
  "status": "FAIL",
  "error": "java.lang.RuntimeException: Request failed with response: {\"Code\":\"InternalError\",\"Message\":\"We encountered an internal error, please try again. (Server not initialized, please try again)\",\"Resource\":\"/minio/admin/v3/add-user\",\"RequestId\":\"1734F24D228B20E8\",\"HostId\":\"0df0ca99-24b7-47c5-9b23-136ff7354b95\"}\n >>> [io.minio.admin.MinioAdminClient.execute(MinioAdminClient.java:206), io.minio.admin.MinioAdminClient.addUser(MinioAdminClient.java:237), TestMinioAdminClient.addUser(TestMinioAdminClient.java:47), TestMinioAdminClient.runAdminTests(TestMinioAdminClient.java:160), FunctionalTest.runTests(FunctionalTest.java:3782), FunctionalTest.runEndpointTests(FunctionalTest.java:3881), FunctionalTest.main(FunctionalTest.java:4002)]"
}
(9/14) Running minio-js tests ... done in 49 seconds
(10/14) Running minio-py tests ... done in 1 minutes and 42 seconds
(11/14) Running s3cmd tests ... done in 15 seconds
(12/14) Running s3select tests ... done in 3 seconds
(13/14) Running versioning tests ... done in 3 minutes and 5 seconds

Executed 13 out of 14 tests successfully.
```
<a name="07e8fc89-1259-4d7c-9e4e-9b31e3229aa4"></a>
### 16323-79822b3/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      15.15.15.6:31259
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp a612c2da04c3:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 1 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 42 seconds
(4/14) Running aws-sdk-ruby tests ... done in 8 seconds
(5/14) Running awscli tests ... done in 59 seconds
(6/14) Running healthcheck tests ... done in 1 seconds
(7/14) Running mc tests ... done in 15 seconds
(8/14) Running minio-go tests ... done in 47 seconds
(9/14) Running minio-java tests ... FAILED in 30 seconds
{
  "name": "minio-java",
  "function": "addUser()",
  "duration": 365,
  "status": "FAIL",
  "error": "java.lang.RuntimeException: Request failed with response: {\"Code\":\"InternalError\",\"Message\":\"We encountered an internal error, please try again. (Server not initialized, please try again)\",\"Resource\":\"/minio/admin/v3/add-user\",\"RequestId\":\"1734F24E58206E1F\",\"HostId\":\"dbc49e30-905a-4a5b-82e1-bb3addea7c4c\"}\n >>> [io.minio.admin.MinioAdminClient.execute(MinioAdminClient.java:206), io.minio.admin.MinioAdminClient.addUser(MinioAdminClient.java:237), TestMinioAdminClient.addUser(TestMinioAdminClient.java:47), TestMinioAdminClient.runAdminTests(TestMinioAdminClient.java:160), FunctionalTest.runTests(FunctionalTest.java:3782), FunctionalTest.runEndpointTests(FunctionalTest.java:3881), FunctionalTest.main(FunctionalTest.java:4002)]"
}
(9/14) Running minio-js tests ... done in 50 seconds
(10/14) Running minio-py tests ... done in 1 minutes and 46 seconds
(11/14) Running s3cmd tests ... done in 17 seconds
(12/14) Running s3select tests ... done in 4 seconds
(13/14) Running versioning tests ... done in 3 minutes and 5 seconds

Executed 13 out of 14 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Review by Praveenrajmani — APPROVED** [review]

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :x: [more...](#0ecb3b11-f84d-4060-a976-568e3ee50100) |
<a name="0ecb3b11-f84d-4060-a976-568e3ee50100"></a>
### 16323-f9be1f5/mint-compress-encrypt-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      15.15.15.6:32479
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp f7b809c005dd:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 1 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 42 seconds
(4/14) Running aws-sdk-ruby tests ... done in 8 seconds
(5/14) Running awscli tests ... done in 1 minutes and 0 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 15 seconds
(8/14) Running minio-go tests ... done in 51 seconds
(9/14) Running minio-java tests ... FAILED in 28 seconds
{
  "name": "minio-java",
  "function": "addUser()",
  "duration": 433,
  "status": "FAIL",
  "error": "java.lang.RuntimeException: Request failed with response: {\"Code\":\"InternalError\",\"Message\":\"We encountered an internal error, please try again. (Server not initialized, please try again)\",\"Resource\":\"/minio/admin/v3/add-user\",\"RequestId\":\"17350A0BBAD38993\",\"HostId\":\"7bdde199-2431-4b15-b3fd-63ac58936fe1\"}\n >>> [io.minio.admin.MinioAdminClient.execute(MinioAdminClient.java:206), io.minio.admin.MinioAdminClient.addUser(MinioAdminClient.java:237), TestMinioAdminClient.addUser(TestMinioAdminClient.java:47), TestMinioAdminClient.runAdminTests(TestMinioAdminClient.java:160), FunctionalTest.runTests(FunctionalTest.java:3782), FunctionalTest.runEndpointTests(FunctionalTest.java:3881), FunctionalTest.main(FunctionalTest.java:4002)]"
}
(9/14) Running minio-js tests ... done in 49 seconds
(10/14) Running minio-py tests ... done in 1 minutes and 41 seconds
(11/14) Running s3cmd tests ... done in 14 seconds
(12/14) Running s3select tests ... done in 4 seconds
(13/14) Running versioning tests ... done in 3 minutes and 4 seconds

Executed 13 out of 14 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Review by poornas — APPROVED** [review]
LGTM

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Pull request merged**